internals/overlord/state/identities.go

```diff
@@ -266,17 +324,43 @@ func (s *State) Identities() map[string]*Identity {
 
 // IdentityFromInputs returns an identity with the given inputs, or nil
 // if there is none.
-func (s *State) IdentityFromInputs(userID *uint32) *Identity {
+//
+// Identity priority:
+//  1. If both username and password are provided, the function attempts to
+//     match a basic type identity. The userID is ignored in this case. If
+//     a matching username is found but the password verification fails, nil
+//     is returned immediately.
+//  2. If username and password are not both provided, the function attempts to
+//     match a local type identity using the userID.
+//
+// If no matching identity is found for the given inputs, nil is returned.
+func (s *State) IdentityFromInputs(userID *uint32, username, password string) *Identity {
 	s.reading()
 
-	for _, identity := range s.identities {
-		switch {
-		case identity.Local != nil && userID != nil:
-			if identity.Local.UserID == *userID {
+	switch {
+	case username != "" || password != "":
```

Comment:
I think we probably want `&&` here (which matches the comment above too - "if *both* username and password are provided").

Reply:
Per Harry's comment, I think we want to be strict and on the safer side: since both user/pass are intentionally set by the user if any is set, it means the user doesn't want to use the "basic" type identity, and it's likely that they used a empty password. If we change it to `&&`, the user might get a basic type identity, if it also has read access, it will be granted metrics access.

For example, if the identities are:

```
"bob": {
	Access: state.ReadAccess,
	Local:  &state.LocalIdentity{UserID: 42},
},
"nancy": {
	Access: state.MetricsAccess,
	Basic:  &state.BasicIdentity{Password: "hash"},
},
```

And if the user input is (42, "nancy", ""), it will match bob although the user's intention is to use nancy since they set the username explicitly.

What do you think?

Reply:
Ah yes, that makes sense. In that case, let's just update the doc comment to match.